### DISCUSSION TITLE
counting ideal points of a hyperbolic line

### DISCUSSION THREAD

**Discussion by mckvv** [cduck/hyperbolic#9] [Q&A] [answered]
Hi, first of all I would like to thank you for your really useful code. I use it to illustrate some of the constructions in the Poicaré disk model :) I am also a beginner in Python programming, so maybe my problem will be trivial, but I would greatly appreciate your help.
My question is, if I have a hyperbolic line defined by two points from inside the disk, how can I calculate its ideal points?

**Comment by cduck** [comment, marked as answer]
I'm glad it's useful to you.  You can find the ideal points of a line by creating a `Line` with `segment=False` (an infinite length line) and asking for its start and end points:
```python
from hyperbolic.poincare import shapes

point1 = shapes.Point.fromEuclid(-0.2, 0.1)
point2 = shapes.Point.fromEuclid(0.2, 0.5)

# Calculate ideal points
line = shapes.Line.fromPoints(*point1, *point2, segment=False)
ideal1 = line.startPoint()  # The ideal point closer to point1
ideal2 = line.endPoint()    # The ideal point closer to point2

# Draw
from hyperbolic import euclid
import drawSvg as draw
d = draw.Drawing(2.1, 2.1, origin='center')
d.draw(euclid.shapes.Circle(0, 0, 1), fill='#eee')
d.draw(line, hwidth=1/8, fill='black')
d.draw(point1, hradius=0.1, fill='green')
d.draw(point2, hradius=0.1, fill='red')
d.draw(ideal1, radius=0.02, fill='lime')
d.draw(ideal2, radius=0.02, fill='orange')

# Display
d.setRenderSize(w=400)
d.savePng('idealPoints.png')
d
```
![idealPoints](https://user-images.githubusercontent.com/2476062/128742326-040ed16b-42a6-41e9-9ecf-614f62b8559b.png)

The calculation that determines the euclidean coordinates of the ideal points for the drawing is [here](https://github.com/cduck/hyperbolic/blob/cd47fbc0bd804af5a4ccc1a2f57950d3c4e76624/hyperbolic/poincare/Line.py#L69).

**Reply by mckvv** [reply]
Hi, only now have I noticed that I didn't answer you. Thank you very much for your help! :)

But I also have another problem. I want to determine the intersection of two hyperbolic line-segments, e.g. AB and CD. But the command AB.intersectionsWithHcycle.CD takes AB and CD as hyperbolic lines. So the result is intersection of the hyperbolic lines AB, CD (if it exists) even if the hyperbolic line-segments do not intersect. Of course, I can investigate wheater the resulting point of intersection is an interior point of these hyperbolic line-segment. But I wonder if there is a more elegant solution :) I also have a problem dealing with this situation if one of those points, e.g. point D is an ideal point (so CD is basically a hyperbolic ray).
Can you help me with that? Am I using the wrong command?

**Reply by cduck** [reply]
Great!

No, there currently is no method that does what you ask.  For new questions like this, it is good practice to open a new discussion or issue.  Would you mind [opening an issue for this feature request](https://github.com/cduck/hyperbolic/issues/new) so I can keep track?  And, if you are interested in contributing, I can show you how to [open a PR](https://github.com/cduck/hyperbolic/pulls) and implement this feature for the package.

**Reply by mckvv** [reply]
Of course, I will open a new issue :) Thank you!

**Reply by cduck** [reply]
#17
